File: frontend/lib/data/services/call_service.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'api_service.dart';
import '../models/call.dart';

class CallService {
  final ApiService _api;

  CallService(this._api);

  Future<List<Call>> getHistory() async {
    final response = await _api.get('/calls/history');
    return (response.data as List).map((e) => Call.fromJson(e)).toList();
  }

  Future<Call> createCall(String receiverId, String type) async {
    final response = await _api.post('/calls/create', data: {
      'receiverId': receiverId,
      'type': type,
    });
    return Call.fromJson(response.data);
  }

  Future<String> getLiveKitToken(String roomName, String participantName) async {
    final response = await _api.post('/calls/livekit-token', data: {
      'roomName': roomName,
      'participantName': participantName,
    });
    return response.data['token'];
  }

  Future<void> updateStatus(String callId, String status, {int? duration}) async {
    await _api.put('/calls/$callId/status', data: {
      'status': status,
      if (duration != null) 'duration': duration,
    });
  }
}

final callServiceProvider = Provider<CallService>((ref) {
  final apiService = ApiService();
  return CallService(apiService);
});

File: backend/src/services/callService.ts
import prisma from '../config/database';
import { generateLiveKitToken } from '../config/livekit';

export class CallService {
  // Получение истории звонков
  static async getHistory(userId: string) {
    return prisma.call.findMany({
      where: {
        OR: [{ callerId: userId }, { receiverId: userId }],
      },
      include: {
        caller: {
          select: {
            id: true,
            firstName: true,
            lastName: true,
            avatar: true,
            username: true,
          },
        },
        receiver: {
          select: {
            id: true,
            firstName: true,
            lastName: true,
            avatar: true,
            username: true,
          },
        },
      },
      orderBy: { createdAt: 'desc' },
      take: 50,
    });
  }

  // Создание звонка
  static async createCall(callerId: string, receiverId: string, type: 'audio' | 'video') {
    const roomName = `call_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;

    const call = await prisma.call.create({
      data: {
        callerId,
        receiverId,
        type,
        status: 'incoming',
        roomName,
      },
      include: {
        caller: {
          select: {
            id: true,
            firstName: true,
            lastName: true,
            avatar: true,
            username: true,
          },
        },
        receiver: {
          select: {
            id: true,
            firstName: true,
            lastName: true,
            avatar: true,
            username: true,
          },
        },
      },
    });

    return call;
  }

  // Получение токена LiveKit
  static async getToken(userId: string, roomName: string, participantName: string) {
    const token = generateLiveKitToken(roomName, participantName, userId);
    return { token };
  }

  // Обновление статуса звонка
  static async updateStatus(callId: string, status: string, duration?: number) {
    return prisma.call.update({
      where: { id: callId },
      data: {
        status,
        ...(duration !== undefined && { duration }),
      },
    });
  }
}
